compiler/rustc_session/src/config/cfg.rs

```diff
@@ -379,3 +380,54 @@ impl CheckCfg {
         ins!(sym::windows, no_values);
     }
 }
+
+pub(crate) fn disallow_cfgs(sess: &Session, user_cfgs: &Cfg) {
+    let disallow = |cfg: &(Symbol, Option<Symbol>), controlled_by| {
+        let cfg_name = cfg.0;
+        let cfg = if let Some(value) = cfg.1 {
+            format!(r#"{}="{}""#, cfg_name, value)
+        } else {
+            format!("{}", cfg_name)
+        };
+        sess.dcx().emit_fatal(DisallowConfig { cfg, cfg_name, controlled_by })
+    };
+
+    // We want to restrict setting cfgs that will produce "incoherent" behavior between
+    // the cfg and the "real" flag that sets it, but not all cfgs produce incoherent
+    // behavior, we therefore exclude those cfgs:
+    //
+    //  - test
+    //  - clippy
+    //  - doc
+    //  - doctest
+    //  - miri
+    //  - rustfmt
+    //  - overflow_checks
+    //  - debug_assertions
+    //  - ub_checks
+
+    for cfg in user_cfgs {
```

Comment:
Can this be merged into `default_configuration` somehow to keep things synchronized?

Reply:
It would probably be better to merge the logic in `fill_well_known` since we know there all the cfgs but `fill_well_known` doesn't do anything when `--check-cfg` isn't active which complicates the logic more; so to keep things simplify for now I would prefer to keep the logic somewhat separate (it's still the same file).

But out of concern I moved the function just above the `default_configuration` function, to hopefully make it more obvious that the function exist and I added the function to the guide (at the top of the file) for when adding a new cfg.